clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -379,6 +382,45 @@ inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
 inlineCleanup _ e = return e
 {-# SCC inlineCleanup #-}
 
+-- | Takes a binding and collapses its term if it is a noop
+inlineIdentity :: HasCallStack => NormRewrite
+inlineIdentity _ (Letrec binds body) = do
+  (untouched,collapsed) <- partitionEithers <$> mapM identityAnalysis binds
+  if null collapsed
+    then return (Letrec binds body)
+    else changed (Letrec (untouched ++ collapsed) body)
+inlineIdentity _ e = return e
+
+identityAnalysis :: HasCallStack => (Id,Term) -> NormalizeSession (Either (Id,Term) (Id,Term))
+identityAnalysis old@(iD,collectArgs -> (term,partitionEithers -> (termArgs,typeArgs))) =
+  runIdentityAnalysis $ do
+    Prim primInfo <- return term
+    identity      <- getIdentity primInfo
+    let resTy = polyResTyView $ primType primInfo
```

Comment:
I don't think this new `polyResTyView` function is what you want. What I think makes more sense is:

  1. Use [`piResultTys`](https://hackage.haskell.org/package/clash-lib-1.4.2/docs/Clash-Core-TermInfo.html#v:piResultTys) to apply the type arguments given to the `Prim` to it's `primType`
  2. Split this type using [`splitFunTys`](https://hackage.haskell.org/package/clash-lib-1.4.2/docs/Clash-Core-Type.html#v:splitFunTys) to get the (now monomorphic) result type and arguments
  3. Index the list of argument types, and use this argument and the result type as type arguments to `unsafeCoerce#`